File: packages/mpv-player/android/src/main/java/expo/modules/mpvplayer/MpvPlayerModule.kt
package expo.modules.mpvplayer

import android.content.Context
import android.hardware.display.DisplayManager
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Build
import android.view.Display
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * The Android twin of `ios/MpvPlayerModule.swift`. Registers the `MpvPlayer` view with the exact same
 * props/events/functions as the Apple module, so the platform-agnostic JS (`requireNativeView("MpvPlayer")`
 * + `use-tv-player`) drives it unchanged. Also exposes the module-level, view-less AUDIO API (§7.14 Phase B /
 * radio channels) via a separate headless `MpvAudioCore`.
 */
class MpvPlayerModule : Module() {
  /** Headless audio-only core, created lazily on first `audioLoad`, independent of the video `View`. */
  private var audio: MpvAudioCore? = null

  override fun definition() = ModuleDefinition {
    Name("MpvPlayer")

    // Module-level (view-less) AUDIO events — the bumper bed + future radio player subscribe to these.
    Events("onAudioProgress", "onAudioEnded", "onAudioError", "onAudioBuffering")

    AsyncFunction("audioLoad") { url: String, startTime: Double -> ensureAudio().load(url, startTime) }
    AsyncFunction("audioAppend") { url: String, startTime: Double -> ensureAudio().append(url, startTime) }
    AsyncFunction("audioPlay") { audio?.play() }
    AsyncFunction("audioPause") { audio?.pause() }
    AsyncFunction("audioStop") { audio?.stop() }
    AsyncFunction("audioSeek") { seconds: Double -> audio?.seek(seconds) }
    AsyncFunction("audioSetVolume") { volume: Double -> audio?.setVolume(volume) }
    AsyncFunction("audioFadeVolume") { volume: Double, durationMs: Double -> audio?.fadeVolume(volume, durationMs) }
    AsyncFunction("audioSetMuted") { muted: Boolean -> audio?.setMuted(muted) }
    AsyncFunction("audioSetRate") { rate: Double -> audio?.setRate(rate) }
    AsyncFunction("audioSetLoop") { loop: Boolean -> audio?.setLoop(loop) }

    // Report what the current audio output can do (Settings → Audio), mirroring the iOS probe. Android has
    // no single "session channels" number, so we take the max channel count across the output devices (the
    // HDMI/eARC sink reports the receiver's real capability) and label the widest one.
    AsyncFunction("getAudioOutputInfo") {
      val am = appContext.reactContext?.applicationContext?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
      var maxCh = 0
      var routeName = ""
      var routeType = ""
      am?.getDevices(AudioManager.GET_DEVICES_OUTPUTS)?.forEach { d ->
        val m = d.channelCounts.maxOrNull() ?: 0
        if (m > maxCh) {
          maxCh = m
          routeName = d.productName?.toString() ?: ""
          routeType = audioDeviceTypeName(d.type)
        }
      }
      mapOf("maxChannels" to maxCh, "currentChannels" to 0, "routeName" to routeName, "routeType" to routeType)
    }

    // The PHYSICAL panel resolution + HDR capability, for Settings → Device. Android TV renders its UI at
    // 1080p on 4K panels, so RN's `Dimensions` (the UI surface) under-reports the panel; the display's
    // supported modes carry the true pixel dimensions. Pattern mirrors jellyfin-androidtv's PlaybackController
    // (getSupportedModes → max physicalWidth×physicalHeight) + deviceProfileReport (HDR types). View-less;
    // uses the default display via DisplayManager (no Activity needed). Synchronous `Function` so the JS
    // device report can read it inline. Android-only — iOS/tvOS keep RN Dimensions (UI surface == panel).
    Function("getDisplayInfo") {
      val ctx = appContext.reactContext?.applicationContext
      val dm = ctx?.getSystemService(Context.DISPLAY_SERVICE) as? DisplayManager
      val display = dm?.getDisplay(Display.DEFAULT_DISPLAY)
      var w = 0
      var h = 0
      display?.supportedModes?.forEach { m ->
        if (m.physicalWidth.toLong() * m.physicalHeight.toLong() > w.toLong() * h.toLong()) {
          w = m.physicalWidth
          h = m.physicalHeight
        }
      }
      val hdrTypes = runCatching {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) display?.mode?.supportedHdrTypes
        else @Suppress("DEPRECATION") display?.hdrCapabilities?.supportedHdrTypes
      }.getOrNull()
      mapOf("width" to w, "height" to h, "hdr" to (hdrTypes?.isNotEmpty() == true))
    }

    OnDestroy {
      audio?.dispose()
      audio = null
    }

    View(MpvPlayerView::class) {
      Events("onLoad", "onFirstFrame", "onProgress", "onBuffering", "onTracks", "onError", "onEnd")

      // `startTime` is set alongside `source` in the same render; store it first, then a source change
      // triggers the coalesced load. (DVR seeks go through the imperative `seek` below, not startTime.)
      Prop("startTime") { view: MpvPlayerView, t: Double -> view.setPendingStartTime(t) }
      Prop("mode") { view: MpvPlayerView, mode: String -> view.setPendingMode(mode) }
      Prop("source") { view: MpvPlayerView, source: String? -> view.setPendingSource(source) }
      Prop("paused") { view: MpvPlayerView, paused: Boolean -> view.setPaused(paused) }
      Prop("muted") { view: MpvPlayerView, muted: Boolean -> view.setMuted(muted) }
      Prop("volume") { view: MpvPlayerView, v: Double -> view.setVolume(v) }
      Prop("contentFit") { view: MpvPlayerView, fit: String -> view.setContentFit(fit) }
      Prop("audioTrack") { view: MpvPlayerView, id: Int -> view.setAudioTrack(id) }
      Prop("subtitleTrack") { view: MpvPlayerView, id: Int -> view.setSubtitleTrack(id) }
      Prop("audioMode") { view: MpvPlayerView, mode: String -> view.setAudioMode(mode) }
      Prop("options") { view: MpvPlayerView, options: Map<String, String>? -> view.options = options ?: emptyMap() }

      AsyncFunction("play") { view: MpvPlayerView -> view.play() }
      AsyncFunction("pause") { view: MpvPlayerView -> view.pause() }
      AsyncFunction("seek") { view: MpvPlayerView, seconds: Double -> view.seek(seconds) }
      // Audio-only controls (bumper music bed + radio) on the same single engine.
      AsyncFunction("fadeVolume") { view: MpvPlayerView, target: Double, durationMs: Double -> view.fadeVolume(target, durationMs) }
      AsyncFunction("setLoop") { view: MpvPlayerView, loop: Boolean -> view.setLoop(loop) }
      AsyncFunction("setRate") { view: MpvPlayerView, rate: Double -> view.setRate(rate) }
      AsyncFunction("append") { view: MpvPlayerView, url: String, startTime: Double -> view.appendTrack(url, startTime) }
      AsyncFunction("setAudioTrack") { view: MpvPlayerView, id: Int -> view.setAudioTrack(id) }
      AsyncFunction("setSubtitleTrack") { view: MpvPlayerView, id: Int -> view.setSubtitleTrack(id) }
    }
  }

  private fun audioDeviceTypeName(type: Int): String = when (type) {
    AudioDeviceInfo.TYPE_HDMI -> "HDMI"
    AudioDeviceInfo.TYPE_HDMI_ARC -> "HDMI ARC"
    AudioDeviceInfo.TYPE_AUX_LINE -> "Line out"
    AudioDeviceInfo.TYPE_BLUETOOTH_A2DP -> "Bluetooth"
    AudioDeviceInfo.TYPE_BUILTIN_SPEAKER -> "Built-in speaker"
    AudioDeviceInfo.TYPE_WIRED_HEADPHONES, AudioDeviceInfo.TYPE_WIRED_HEADSET -> "Wired"
    else -> "Output"
  }

  private fun ensureAudio(): MpvAudioCore {
    audio?.let { return it }
    // Audio needs no Activity/UI — the application context is enough (mirrors plezy's audio plugin).
    val ctx = appContext.reactContext?.applicationContext ?: appContext.reactContext!!
    val core = MpvAudioCore(ctx)
    core.onProgress = { time, duration ->
      sendEvent("onAudioProgress", mapOf("currentTime" to time, "duration" to duration))
    }
    core.onEnded = { sendEvent("onAudioEnded", emptyMap<String, Any>()) }
    core.onError = { message -> sendEvent("onAudioError", mapOf("message" to message)) }
    core.onBuffering = { buffering -> sendEvent("onAudioBuffering", mapOf("buffering" to buffering)) }
    core.setup()
    audio = core
    return core
  }
}

File: packages/mpv-player/ios/MpvPlayerModule.swift
import AVFoundation
import ExpoModulesCore

public class MpvPlayerModule: Module {
  /// Headless audio-only core (§7.14 Phase B / radio channels). Created lazily on first `audioLoad`, held for
  /// the module's lifetime, and independent of the video `View`. See `MpvAudioCore`.
  private var audio: MpvAudioCore?

  public func definition() -> ModuleDefinition {
    Name("MpvPlayer")

    // Configure the app's ONE shared AVAudioSession at module init (app launch) — plezy's pattern. Both
    // mpv cores (video + headless audio) then just PLAY THROUGH it and never touch it themselves, so
    // neither the bumper-music core's start nor its teardown can flip the session out from under the video
    // (the post-bumper re-pause + lost-5.1). `.longFormAudio` negotiates the full multichannel route; and a
    // fresh tune that lands straight on a bumper still has an active session for the music bed.
    OnCreate {
      let session = AVAudioSession.sharedInstance()
      try? session.setCategory(.playback, mode: .moviePlayback, policy: .longFormAudio, options: [])
      try? session.setActive(true)
    }

    // Module-level (view-less) AUDIO events — the bumper bed + future radio player subscribe to these.
    Events("onAudioProgress", "onAudioEnded", "onAudioError", "onAudioBuffering")

    AsyncFunction("audioLoad") { (url: String, startTime: Double) in self.ensureAudio().load(url, startTime: startTime) }
    AsyncFunction("audioAppend") { (url: String, startTime: Double) in self.ensureAudio().append(url, startTime: startTime) }
    AsyncFunction("audioPlay") { self.audio?.play() }
    AsyncFunction("audioPause") { self.audio?.pause() }
    AsyncFunction("audioStop") { self.audio?.stop() }
    AsyncFunction("audioSeek") { (seconds: Double) in self.audio?.seek(seconds) }
    AsyncFunction("audioSetVolume") { (volume: Double) in self.audio?.setVolume(volume) }
    AsyncFunction("audioFadeVolume") { (volume: Double, durationMs: Double) in self.audio?.fadeVolume(to: volume, durationMs: durationMs) }
    AsyncFunction("audioSetMuted") { (muted: Bool) in self.audio?.setMuted(muted) }
    AsyncFunction("audioSetRate") { (rate: Double) in self.audio?.setRate(rate) }
    AsyncFunction("audioSetLoop") { (loop: Bool) in self.audio?.setLoop(loop) }

    // Report what the current audio output route advertises, so the app can show whether real 5.1/7.1 is
    // available (Settings → Audio). Configures the same long-form playback session the video player uses so
    // the numbers reflect what playback would actually get (idempotent — safe while audio is playing).
    AsyncFunction("getAudioOutputInfo") { () -> [String: Any] in
      let session = AVAudioSession.sharedInstance()
      try? session.setCategory(.playback, mode: .moviePlayback, policy: .longFormAudio, options: [])
      try? session.setActive(true)
      let out = session.currentRoute.outputs.first
      return [
        "maxChannels": session.maximumOutputNumberOfChannels,
        "currentChannels": session.outputNumberOfChannels,
        "routeName": out?.portName ?? "",
        "routeType": out?.portType.rawValue ?? "",
      ]
    }

    OnDestroy {
      self.audio?.dispose()
      self.audio = nil
    }

    View(MpvPlayerView.self) {
      Events("onLoad", "onFirstFrame", "onProgress", "onBuffering", "onTracks", "onError", "onEnd")

      // `startTime` is set alongside `source` in the same render; store it first, then a source change
      // triggers the coalesced load. (DVR seeks go through the imperative `seek` below, not startTime.)
      Prop("startTime") { (view: MpvPlayerView, t: Double) in view.setPendingStartTime(t) }
      Prop("mode") { (view: MpvPlayerView, mode: String) in view.setPendingMode(mode) }
      Prop("source") { (view: MpvPlayerView, source: String?) in view.setPendingSource(source) }
      Prop("paused") { (view: MpvPlayerView, paused: Bool) in view.setPaused(paused) }
      Prop("muted") { (view: MpvPlayerView, muted: Bool) in view.setMuted(muted) }
      Prop("volume") { (view: MpvPlayerView, v: Double) in view.setVolume(v) }
      Prop("contentFit") { (view: MpvPlayerView, fit: String) in view.setContentFit(fit) }
      Prop("audioTrack") { (view: MpvPlayerView, id: Int) in view.setAudioTrack(id) }
      Prop("subtitleTrack") { (view: MpvPlayerView, id: Int) in view.setSubtitleTrack(id) }
      Prop("audioMode") { (view: MpvPlayerView, mode: String) in view.setAudioMode(mode) }
      Prop("options") { (view: MpvPlayerView, options: [String: String]?) in view.options = options ?? [:] }

      AsyncFunction("play") { (view: MpvPlayerView) in view.play() }
      AsyncFunction("pause") { (view: MpvPlayerView) in view.pause() }
      AsyncFunction("seek") { (view: MpvPlayerView, seconds: Double) in view.seek(seconds) }
      // Audio-only controls (bumper music bed + radio) on the same single engine.
      AsyncFunction("fadeVolume") { (view: MpvPlayerView, target: Double, durationMs: Double) in view.fadeVolume(target, durationMs) }
      AsyncFunction("setLoop") { (view: MpvPlayerView, loop: Bool) in view.setLoop(loop) }
      AsyncFunction("setRate") { (view: MpvPlayerView, rate: Double) in view.setRate(rate) }
      AsyncFunction("append") { (view: MpvPlayerView, url: String, startTime: Double) in view.appendTrack(url, startTime) }
      AsyncFunction("setAudioTrack") { (view: MpvPlayerView, id: Int) in view.setAudioTrack(id) }
      AsyncFunction("setSubtitleTrack") { (view: MpvPlayerView, id: Int) in view.setSubtitleTrack(id) }
    }
  }

  private func ensureAudio() -> MpvAudioCore {
    if let audio { return audio }
    let core = MpvAudioCore()
    core.onProgress = { [weak self] time, duration in
      self?.sendEvent("onAudioProgress", ["currentTime": time, "duration": duration])
    }
    core.onEnded = { [weak self] in self?.sendEvent("onAudioEnded", [:]) }
    core.onError = { [weak self] message in self?.sendEvent("onAudioError", ["message": message]) }
    core.onBuffering = { [weak self] buffering in self?.sendEvent("onAudioBuffering", ["buffering": buffering]) }
    core.setup()
    audio = core
    return core
  }
}
